### PR TITLE
refator(DialogGuidanceFeedbackRuleEvaluator): Extract TermEvaluators

### LINKED ISSUE #763: refactor(DGFeedbackRuleEvaluator): Extract TermEvaluators
### Description

We have 3 term evaluators that are in the DialogGuidanceFeedbackRuleEvaluator class, which can be extracted into their own class:
- HasKIScore
- IdeaCount[MoreThan|Equals|LessThan]
- Idea

### PR DESCRIPTION
**Pull request by hirokiterashima**
## Changes
- Extract 3 term evaluators from DialogGuidanceFeedbackRuleEvaluator into their own class
   - HasKIScore => HasKIScoreTermEvaluator
   - IdeaCount[MoreThan|Equals|LessThan] => IdeaCountTermEvaluator
   - Idea => IdeaTermEvaluator
- Each evaluator extends abstract TermEvaluator class
- No new feature was added

## Test
- Dialog Guidance feedback works as before
   - single score, multiple score
   - rules: hasKIScore(X), ideaCount[MoreThan|Equals|LessThan](X), ideaX 

Closes #763

### CODE REVIEW

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaCountTermEvaluator.spec.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+const evaluator_more_than_2 = new IdeaCountTermEvaluator('ideaCountMoreThan(2)');
+const evaluator_equals_2 = new IdeaCountTermEvaluator('ideaCountEquals(2)');
+const evaluator_less_than_2 = new IdeaCountTermEvaluator('ideaCountLessThan(2)');
+describe('IdeaCountTermEvaluator', () => {
+  describe('evaluate()', () => {
+    it('should check whether number of detected ideas meets criteria', () => {
+      const idea1 = new CRaterIdea('idea1', true);
+      const idea2 = new CRaterIdea('idea2', true);
+      const idea3 = new CRaterIdea('idea3', true);
```
Comment (geoffreykwan) [inline src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaCountTermEvaluator.spec.ts]: Should the parameter to CRaterIdea not include the 'idea' text to more accurately represent how it's actually used when students have their work scored?
```
      const idea1 = new CRaterIdea('1', true);
      const idea2 = new CRaterIdea('2', true);
      const idea3 = new CRaterIdea('3', true);
```

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import { CRaterIdea } from '../CRaterIdea';
+import { CRaterResponse } from '../CRaterResponse';
+import { IdeaTermEvaluator } from './IdeaTermEvaluator';
+
+describe('IdeaTermEvaluator', () => {
+  describe('evaluate()', () => {
+    const response_with_no_ideas = new CRaterResponse();
+    const response_with_2_ideas = new CRaterResponse();
+    response_with_2_ideas.ideas = [new CRaterIdea('idea1', true), new CRaterIdea('idea2', true)];
```
Comment (geoffreykwan) [inline src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts]: Should the parameter to CRaterIdea not include the 'idea' text?
```
    response_with_2_ideas.ideas = [new CRaterIdea('1', true), new CRaterIdea('2', true)];
```

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    it("should always return true if term is 'true'", () => {
+      const evaluator_true = new IdeaTermEvaluator('true');
+      expect(evaluator_true.evaluate(response_with_no_ideas)).toBeTrue();
+      expect(evaluator_true.evaluate(response_with_2_ideas)).toBeTrue();
+    });
+    it('should return true iff idea term is found in the response', () => {
+      const evaluator_idea1 = new IdeaTermEvaluator('idea1');
```
Comment (geoffreykwan) [inline src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts]: Should the parameter to IdeaTermEvaluator not include the 'idea' text?
```
      const evaluator_idea1 = new IdeaTermEvaluator('1');
```

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { CRaterResponse } from '../CRaterResponse';
+import { TermEvaluator } from './TermEvaluator';
+
+export class IdeaTermEvaluator extends TermEvaluator {
+  constructor(protected term: string) {
+    super(term);
+  }
+
+  evaluate(response: CRaterResponse): boolean {
+    return this.term === 'true' || response.getDetectedIdeaNames().includes(this.term);
```
Comment (geoffreykwan) [inline src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts:18]: What is the purpose of ```this.term === 'true'```? Do authors actually use the true expression?

**Review by geoffreykwan — CHANGES_REQUESTED** [review]
Works well. I have some test change suggestions and a question.

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts (hunk 4 above)
Comment (hirokiterashima) [reply src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts:18]: The author would usually not use the true expression when writing rules (but they can if they want to).

The term 'true' or 'false' is created and added to the termStack by the application as it processes operator expressions like "&&", "||", and "!" (see ```DialogGuidanceFeedbackRuleEvaluator.evaluateOperator()```). 

If student had ideas 1, 2 and 3, evaluating the postfix expression ["1", "2", "&&", "3", "&&"] goes like this:
["1", "2", "&&", "3", "&&"] => ["true", "3", "&&"] => ["true"] => true.

I agree this code is confusing at a glance. For now, I can add a comment on the function about why we need this check, and we can think of a better way to write this in the future?

Comment (geoffreykwan) [reply src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts:18]: Yea a comment would be fine.

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts (hunk 2 above)
Comment (hirokiterashima) [reply src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts]: done

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts (hunk 4 above)
Comment (hirokiterashima) [reply src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.ts:18]: Comment added.

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts (hunk 3 above)
Comment (hirokiterashima) [reply src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaTermEvaluator.spec.ts]: done

File: src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaCountTermEvaluator.spec.ts (hunk 1 above)
Comment (hirokiterashima) [reply src/assets/wise5/components/dialogGuidance/TermEvaluator/IdeaCountTermEvaluator.spec.ts]: done

**Review by geoffreykwan — APPROVED** [review]
Looks good.

**Pull request merged**